### ISSUE TITLE
`ValueConverter` converts first custom rule and ignores the rest custom ones

### ISSUE THREAD

**Issue by Yokozuna59** [langium/langium#1071]
Continuing from conversion <https://github.com/langium/langium/pull/1067#issuecomment-1574258361>.

Langium version: 1.2
Package name: mermaid-parser

## Steps To Reproduce

1. Create custom `ValueConverter` by extending the default.
2. Set custom pattern function.
3. Add Value Converter to language module.
4. Add the custom `TokenBuilder` to language module.
5. Parse the code.

Link to code example:

- Custom `ValueConverter`: <https://github.com/Yokozuna59/mermaid-lint/blob/44e3428e94087a4e82257767403c92726590c8f3/packages/mermaid-parser/src/language/mermaid-value-converter.ts#L13-L51>
- Custom pattern function: <https://github.com/Yokozuna59/mermaid-lint/blob/44e3428e94087a4e82257767403c92726590c8f3/packages/mermaid-parser/src/language/matchers/common.ts#L29-L36>
- Custom `TokenBuilder`: <https://github.com/Yokozuna59/mermaid-lint/blob/44e3428e94087a4e82257767403c92726590c8f3/packages/mermaid-parser/src/language/mermaid-token-builder.ts#L11-L37>
- Sample code to parse: <https://github.com/Yokozuna59/mermaid-lint/blob/44e3428e94087a4e82257767403c92726590c8f3/packages/mermaid-parser/tests/common/title.test.ts#L146-L158>

## The current behavior

The entered input:

```
pie title test
title sample
```

Enters matcher functions:
 - First match:
  ![image](https://github.com/langium/langium/assets/87622592/954a0bda-d660-4e98-93c9-609e40b0ce59)
  
- Second match:
  ![image](https://github.com/langium/langium/assets/87622592/5f44798f-b346-4446-850e-614a857c408f)

Enters `ValueConverter`

I added `console.log` in `ValueConverter` to see how many times have been entered, only once:

![image](https://github.com/langium/langium/assets/87622592/25b25124-c638-4c2c-ab15-1dabb85a13fc)

Then goes stray to building AST.

> **Note**: It's occurs when there is more than one value need to be converted, not only duplicate rule.

## The expected behavior

Matches both custom rules, convert values, set the value of both with no errors.

**Comment by msujew** [comment]
@Yokozuna59 again, the issue is somewhere else. Your fragment rule `TITLE_AND_ACCESSIBILITIES` isn't parsing `NEWLINE` even though it's in the input between both `TITLE` terminals. Replacing `TITLE_AND_ACCESSIBILITIES` rule with the following does the trick:

```
fragment TITLE_AND_ACCESSIBILITIES:
    ((accDescr=ACC_DESCR | accTitle=ACC_TITLE | title=TITLE) NEWLINE*)+
;
```

**Comment by Yokozuna59** [comment]
It works, but that would break the `Section` parse rule that needs `NEWLINE+` if `TITLE_AND_ACCESSIBILITIES` fragment exists.

**Comment by msujew** [comment]
You can probably move `NEWLINE*` in front of the assignments instead of after it, that should fix it.

**Comment by Yokozuna59** [comment]
> You can probably move `NEWLINE*` in front of the assignments instead of after it, that should fix it.

No it didn't, it also brook other parts other than section test cases. Here is the parser errors:

```
[
  NoViableAltException: Expecting: one of these possible Token sequences:
    1. [ACC_DESCR]
    2. [ACC_TITLE]
    3. [TITLE]
  but found: '"GitHub"'
      at ChevrotainWrapper.ErrorHandler.raiseNoAltException (/home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/chevrotain@10.4.2/node_modules/chevrotain/src/parse/parser/traits/error_handler.ts:120:7)
      at ChevrotainWrapper.RecognizerEngine.orInternal (/home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/chevrotain@10.4.2/node_modules/chevrotain/src/parse/parser/traits/recognizer_engine.ts:652:10)
      at ChevrotainWrapper.RecognizerApi.or (/home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/chevrotain@10.4.2/node_modules/chevrotain/src/parse/parser/traits/recognizer_api.ts:71:17)
      at ChevrotainWrapper.wrapOr (/home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/langium@1.2.0/node_modules/langium/src/parser/langium-parser.ts:559:14)
      at LangiumParser.alternatives (/home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/langium@1.2.0/node_modules/langium/src/parser/langium-parser.ts:83:22)
      at /home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/langium@1.2.0/node_modules/langium/src/parser/parser-builder-base.ts:176:37
      at /home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/langium@1.2.0/node_modules/langium/src/parser/parser-builder-base.ts:255:48
      at Array.forEach (<anonymous>)
      at /home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/langium@1.2.0/node_modules/langium/src/parser/parser-builder-base.ts:255:30
      at ChevrotainWrapper.DEF (/home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/langium@1.2.0/node_modules/langium/src/parser/parser-builder-base.ts:351:28) {
    token: {
      image: '"GitHub"',
      startOffset: 41,
      endOffset: 48,
      startLine: 2,
      endLine: 2,
      startColumn: 17,
      endColumn: 24,
      tokenTypeIdx: 16,
      tokenType: [Object]
    },
    resyncedTokens: [],
    previousToken: {
      image: '\n',
      startOffset: 24,
      endOffset: 24,
      startLine: 1,
      endLine: 1,
      startColumn: 25,
      endColumn: 25,
      tokenTypeIdx: 8,
      tokenType: [Object]
    },
    context: { ruleStack: [Array], ruleOccurrenceStack: [Array] }
  },
  NotAllInputParsedException: Expecting end of file but found `"GitHub"`.
      at ChevrotainWrapper.RecognizerEngine.ruleFinallyStateUpdate (/home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/chevrotain@10.4.2/node_modules/chevrotain/src/parse/parser/traits/recognizer_engine.ts:671:23)
      at ChevrotainWrapper.invokeRuleWithTryCst [as Mermaid​] (/home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/chevrotain@10.4.2/node_modules/chevrotain/src/parse/parser/traits/recognizer_engine.ts:245:16)
      at LangiumParser.parse (/home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/langium@1.2.0/node_modules/langium/src/parser/langium-parser.ts:154:38)
      at DefaultLangiumDocumentFactory.parse (/home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/langium@1.2.0/node_modules/langium/src/workspace/documents.ts:251:46)
      at DefaultLangiumDocumentFactory.create (/home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/langium@1.2.0/node_modules/langium/src/workspace/documents.ts:170:38)
      at DefaultLangiumDocumentFactory.fromString (/home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/langium@1.2.0/node_modules/langium/src/workspace/documents.ts:156:21)
      at parse (/home/potato/work/yokozuna59/mermaid-lint/packages/mermaid-parser/tests/test-utils.ts:24:59)
      at /home/potato/work/yokozuna59/mermaid-lint/packages/mermaid-parser/tests/pie/syntax.test.ts:274:49
      at runTest (file:///home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/@vitest+runner@0.31.4/node_modules/@vitest/runner/dist/index.js:508:11)
      at runSuite (file:///home/potato/work/yokozuna59/mermaid-lint/node_modules/.pnpm/@vitest+runner@0.31.4/node_modules/@vitest/runner/dist/index.js:611:15) {
    token: {
      image: '"GitHub"',
      startOffset: 41,
      endOffset: 48,
      startLine: 2,
      endLine: 2,
      startColumn: 17,
      endColumn: 24,
      tokenTypeIdx: 16,
      tokenType: [Object]
    },
    resyncedTokens: [],
    context: { ruleStack: [], ruleOccurrenceStack: [] }
  }
]```

**Comment by msujew** [comment]
@Yokozuna59 That's an issue you encounter due to `NEWLINE` tokens at the start of groups. The parser will enter the optional group due to it's lookahead matching a `NEWLINE` token (or multiple). Since it's already inside of the group, it will try to parse `TITLE_AND_ACCESSIBILITIES` as well, which has a mandatory `ACC_DESCR|ACC_TITLE|TITLE` group in there. That's the reason for the parser error.

Again, I would wholeheartedly recommend to rethink your approach to parsing `NEWLINE` and move it to the end of groups/statements. Having them at the start of groups will always lead to problems somewhere down the line.

**Comment by Yokozuna59** [comment]
> @Yokozuna59 That's an issue you encounter due to `NEWLINE` tokens at the start of groups. The parser will enter the optional group due to it's lookahead matching a `NEWLINE` token (or multiple). Since it's already inside of the group, it will try to parse `TITLE_AND_ACCESSIBILITIES` as well, which has a mandatory `ACC_DESCR|ACC_TITLE|TITLE` group in there. That's the reason for the parser error.

Oh, I see why. It appears that it would be hard to implement that fragment rule so that it doesn't affect other parts that needs `NEWLINE`.

> Again, I would wholeheartedly recommend to rethink your approach to parsing `NEWLINE` and move it to the end of groups/statements. Having them at the start of groups will always lead to problems somewhere down the line.

It would be really hard to implement it in this way; some parts need to have `NEWLINE` at the beginning, and some don't.

For instance, in the `Section` parse rule, the first section would be parsed currently if `TITLE_AND_ACCESSIBILITIES` exists since that fragment would parse the whole line, but what about if it doesn't? then the following code would be fine:

```
pie "section": 1
```

And I can't add `NEWLINE+` after `Section` because it'll throw errors if it was at `EOF`. The solution here didn't work https://github.com/langium/langium/discussions/1048#discussioncomment-5948913.

---

Anyway, I'll close the issue since it appears that the issue was on the grammar side, not the langium side as I thought. I really apologize for disturbing you guys the past few weeks because of my lack of skills in the great framework.

Thanks for your help so much!

**Comment by Yokozuna59** [comment]
I might do what you have suggested:

> 2. Override the `Lexer` service to always return a `NEWLINE` token at the end of the token stream. That way you don't need to work with the `EOF` token (which doesn't seem to work, we've also experienced this issue recently).

To add `NEWLINE` token after each time match of `TITLE_AND_ACCESSIBILITIES` fragment rule to prevent greedily consume of newlines.

**Comment by msujew** [comment]
> I really apologize for disturbing you guys the past few weeks because of my lack of skills in the great framework.

All good, you're asking questions in good faith to learn something new. We've all been there :)

I am by the way very interested in seeing a Langium based parser merged into the Mermaid code base. This is what you're discussing in the other issue, right? This would be a huge boost in download numbers and popularity for Langium.

> To add NEWLINE token after each time match of TITLE_AND_ACCESSIBILITIES fragment rule to prevent greedily consume of newlines. But not sure it if would affect other parts from location perspective.

Yes, that's one way to do it. Looking at the original [jison grammar](https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/diagrams/pie/parser/pie.jison), I begin to wonder why you don't transcribe it one to one into the Langium grammar? The jison grammar also seems to use `eol` at the end of each statement. They're both EBNF-esque language definitions, so that should work without any issues. Changing the grammar like you're doing right now will certainly result in edge case regressions from previous mermaid versions.

**Comment by Yokozuna59** [comment]
> All good, you're asking questions in good faith to learn something new. We've all been there :)

Thanks!

> I am by the way very interested in seeing a Langium based parser merged into the Mermaid code base. This is what you're discussing in the other issue, right?

Yes! That's the goal; I'm currently planning to support it as a standalone parser before supporting other services, e.g., LSP, etc.

I'm still not sure what's the best approach for using it as parser not a LSP.

Here is what I come up with in this [pr](https://github.com/mermaid-js/mermaid/blob/8caaf7c36ca7741ecf7baf95861fb599b86e6831/packages/mermaid/src/diagrams/pie/pieParser.ts#L7-L29).

> Looking at the original [jison grammar](https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/diagrams/pie/parser/pie.jison), I begin to wonder why you don't transcribe it one to one into the Langium grammar? The jison grammar also seems to use `eol` at the end of each statement. They're both EBNF-esque language definitions, so that should work without any issues. Changing the grammar like you're doing right now will certainly result in edge case regressions from previous mermaid versions.

As I said before, the `EOF` token from `chevrotain` isn't working as expected, or maybe I used it in the wrong way, and it's properly the second one.

Maybe because it uses it as [/NOT_APPLICABLE/](https://github.com/Chevrotain/chevrotain/blob/6d210a382b6473379aa7e40857ecf42e14b1248f/packages/chevrotain/src/scan/lexer_public.ts#L93) instead of really `EOF`.

**Comment by msujew** [comment]
> As I said before, the EOF token from chevrotain isn't working as expected, or maybe I used it in the wrong way, and it's properly the second one.

Yeah, I'm not sure how to deal with `EOF` in Chevrotain either. My recommendation would be to add a `NEWLINE` token to the end of the token stream in the `Lexer` service. That would alleviate the need for trying to parse the `EOF` token at all. See [here](https://github.com/BBx-Kitchen/bbj-language-server/blob/8b6c6d72a6e5f62ac52a851aeb1db18a933671a0/bbj-vscode/src/language/bbj-lexer.ts) for an example.
